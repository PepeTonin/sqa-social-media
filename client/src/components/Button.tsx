import React from "react";

interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: "primary" | "secondary" | "outline";
  isLoading?: boolean;
}

export default function Button({
  children,
  variant = "primary",
  isLoading = false,
  disabled,
  ...props
}: ButtonProps) {
  const getVariantStyles = () => {
    switch (variant) {
      case "primary":
        return {
          background: "var(--primary)",
          color: "white",
          border: "none",
        };
      case "secondary":
        return {
          background: "var(--secondary)",
          color: "white",
          border: "none",
        };
      case "outline":
        return {
          background: "transparent",
          color: "var(--foreground)",
          border: "1px solid var(--border)",
        };
      default:
        return {};
    }
  };

  const getHoverColor = () => {
    switch (variant) {
      case "primary":
        return "var(--primary-hover)";
      case "secondary":
        return "var(--secondary-hover)";
      case "outline":
        return "var(--card-hover)";
      default:
        return "";
    }
  };

  return (
    <button
      {...props}
      disabled={disabled || isLoading}
      style={{
        ...getVariantStyles(),
        padding: "0.75rem 1.5rem",
        borderRadius: "0.375rem",
        cursor: disabled || isLoading ? "not-allowed" : "pointer",
        fontWeight: "500",
        fontSize: "1rem",
        transition: "all 0.2s",
        opacity: disabled || isLoading ? 0.7 : 1,
        ...props.style,
      }}
      onMouseOver={(e) => {
        if (!disabled && !isLoading && variant !== "outline") {
          e.currentTarget.style.background = getHoverColor();
        } else if (!disabled && !isLoading && variant === "outline") {
          e.currentTarget.style.background = getHoverColor();
        }
      }}
      onMouseOut={(e) => {
        const background = getVariantStyles().background ?? "";
        if (variant !== "outline" && background) {
          e.currentTarget.style.background = background;
        } else {
          e.currentTarget.style.background = "transparent";
        }
      }}
    >
      {isLoading ? "Carregando..." : children}
    </button>
  );
}
