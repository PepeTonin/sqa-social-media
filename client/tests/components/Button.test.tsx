import { fireEvent, render } from "@testing-library/react";
import Button from "@/components/Button";

describe("Button", () => {
  describe("function renderButton", () => {
    test("deve gerar o snapshot e ser idêntico ao snapshot anterior", () => {
      const { asFragment } = render(
        <Button onClick={() => {}}>Click me</Button>
      );
      expect(asFragment()).toMatchSnapshot();
    });

    test("deve renderizar o botão com o texto 'Click me'", () => {
      const { getByText } = render(
        <Button onClick={() => {}}>Click me</Button>
      );
      expect(getByText("Click me")).toBeDefined();
    });

    test("deve chamar a função onClick quando o botão for clicado", () => {
      const onClick = jest.fn();
      const { getByText } = render(<Button onClick={onClick}>Click me</Button>);
      fireEvent.click(getByText("Click me"));
      expect(onClick).toHaveBeenCalled();
    });
  });
});
