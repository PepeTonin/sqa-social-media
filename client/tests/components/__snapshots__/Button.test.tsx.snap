// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Button function renderButton deve gerar o snapshot e ser idÃªntico ao snapshot anterior 1`] = `
<DocumentFragment>
  <button
    style="background: var(--primary); color: white; padding: 0.75rem 1.5rem; border-radius: 0.375rem; cursor: pointer; font-weight: 500; font-size: 1rem; transition: all 0.2s; opacity: 1;"
  >
    Click me
  </button>
</DocumentFragment>
`;
